compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -503,6 +503,14 @@ where
                 // Don't go into the same opaque type again.
                 if ty != self.skip {
```

Comment:
it feels weird to only skip the innermost opaque type :thinking: are there no cases where that can go wrong?

e.g. can you have something like
```rust
type Foo = impl Trait<Assoc = Bar>;
type Bar = impl Trait<Assoc = FoO>;
```

Reply:
that's an expected cycle I think. The reason this works is that `type Foo = impl Trait<Assoc = XXX>` desugars to two predicates:

* `Foo: Trait`
* `<Foo as Trait>::Assoc = XXX`

And we know this always the desugaring that happens. It may be surprising once users write `where Foo: Bop` bounds on type aliases, but I think it doesn't matter whether they write explicit where bounds or use the impl trait "sugar".